x-pack/plugins/security_solution/server/usage/detections/detections_helpers.ts

```diff
@@ -301,3 +483,116 @@ export const getMlJobMetrics = async (
 
   return [];
 };
+
+export const getDetectionRuleMetrics = async (
+  kibanaIndex: string,
+  signalsIndex: string,
+  esClient: ElasticsearchClient,
+  savedObjectClient: SavedObjectsClientContract
+): Promise<DetectionRuleAdoption> => {
+  let rulesUsage: DetectionRulesTypeUsage = initialDetectionRulesUsage;
+  const ruleSearchOptions: RuleSearchParams = {
+    body: { query: { bool: { filter: { term: { 'alert.alertTypeId': SIGNALS_ID } } } } },
+    filterPath: [],
+    ignoreUnavailable: true,
+    index: kibanaIndex,
+    size: 10_000,
+  };
+
+  let detectionAlertsResp: AlertsAggregationResponse | undefined;
+
+  try {
+    const { body: ruleResults } = await esClient.search<RuleSearchResult>(ruleSearchOptions);
+
+    // @ts-expect-error `SearchResponse['hits']['total']` incorrectly expects `number` type instead of `{ value: number }`.
+    ({ body: detectionAlertsResp } = await esClient.search({
+      index: signalsIndex,
+      size: 0,
+      body: {
+        aggs: {
+          detectionAlerts: {
+            terms: { field: 'signal.rule.id.keyword' },
+          },
+        },
+        query: {
+          bool: {
+            filter: [
+              {
+                range: {
+                  '@timestamp': {
+                    gte: 'now-24h',
+                    lte: 'now',
+                  },
+                },
+              },
+            ],
+          },
+        },
+      },
+    }));
+
+    const cases = await savedObjectClient.find<CasesSavedObject>({
+      type: 'cases-comments',
+      fields: [],
+      page: 1,
+      perPage: 10_000,
+      filter: 'cases-comments.attributes.type: alert',
+    });
+
+    const casesCache = cases.saved_objects.reduce((cache, { attributes: casesObject }) => {
+      const ruleId = casesObject.rule.id;
+
+      const cacheCount = cache.get(ruleId);
+      if (cacheCount === undefined) {
+        cache.set(ruleId, 1);
+      } else {
+        cache.set(ruleId, cacheCount + 1);
+      }
+      return cache;
+    }, new Map<string, number>());
+
+    const alertBuckets = detectionAlertsResp!.aggregations?.detectionAlerts?.buckets ?? [];
+
+    const alertsCache = new Map();
+    alertBuckets.map((bucket) => alertsCache.set(bucket.key, bucket.doc_count));
+
+    if (ruleResults.hits?.hits?.length > 0) {
+      const ruleObjects = ruleResults.hits?.hits?.map((hit) => {
+        const ruleId = hit._id.split(':')[1];
+        // @ts-expect-error _source is optional
+        const isElastic = isElasticRule(hit._source?.alert.tags);
```

Comment:
Instead of all these `ts-expect-error` which turn off the type checking completely and most think is a code smell, can we just change `isElasticRule` to default to an empty array if it is passed an undefined like this?

```ts
const isElasticRule = (tags: string[] = []) => tags.includes(`${INTERNAL_IMMUTABLE_KEY}:true`);
```

Then we can remove all the `ts-expect-error` around the `hit._source`.

If we do this and that as cast above, this whole file would be free of `ts-expect-error` and that would be super awesome!

Reply:
Thanks! Updated in 0c91a78
I learnt a lot from your comments! Thanks for the review!